Documentation for parameterized client scopes (#50037)
keycloak/keycloak: docs/documentation/server_admin/topics/clients/con-parameterized-client-scopes.adoc

=== Creating a parameterized client scope

To create a parameterized client scope, follow these steps:

. Click *Client scopes* in the menu.
. Click *Create client scope*.
. Enter a *Name* for the scope (for example, `project`).
. Enable the *Parameterized scope* toggle.
. Select a *Parameter type* from the available options (for example, `String`).
. If you selected the `Custom` parameter type, enter a regular expression pattern that the parameter value must match.
. Set *Type* to *None* or *Optional*. When the *Parameterized scope* toggle is enabled, the *Default* option is not available.
. Click *Save*.

.Creating a parameterized client scope
image:images/client-scopes-parameterized-create.png[width=600]

=== Parameter types

When creating a parameterized scope, you must select a parameter type that determines how the parameter value is validated at request time. The following built-in types are available:

[cols="1,3",options="header"]
|===
| Type | Description

| `string`
| Accepts any string value. No validation is performed on the parameter.

| `integer`

> | Accepts whole numbers only. The parameter value must be parsable as a number.

Reviewer: A whole number is non-negative integer number, so I would use just integer here as we are accepting negative numbers...

```suggestion
| Accepts integer numbers only. The parameter value must be parsable as an integer number.
```

| `boolean`
| Accepts only `true` or `false` (case-insensitive).

| `username`

> | Validates that the parameter value is the username of an existing, enabled user in the realm. Additionally, a user cannot target themselves with this scope type.

Author: @rmartinc Just wondering. 

Right now, we don't allow targeting the user themselves. Do we have any use case, when would be required to target themselves? 

Additionally, when the user is not found, or is disabled, we just throw an error with the appropriate message - it provides some sort of user enumeration. Is it ok for us?

Reviewer: Yep, I remember that, I really don't know if a use-case can accept the user himself. Let this for the moment like it is now.

About the user enumeration, yes, we are returning invalid scopes. But this is what we are doing for any other scope. I suppose we can silently remove the scope but that means a change for normal scopes. Maybe only doing that for parameterized or even for the username type. But I would maintain this for the moment.
